Tasks/DeployAzureResourceGroup/MachineGroupAgentExtensionManager.ts

```diff
@@ -0,0 +1,294 @@
+            operationParameters.deferred = Q.defer<string>();
+            var listOfVmsPromise = this._azureUtils.getVMDetails();
+            listOfVmsPromise.then((listOfVms) => {
+                operationParameters.vmCount = listOfVms.length;
+                if (operationParameters.vmCount == 0) {
+                    operationParameters.deferred.resolve("");
+                }
+                for (var i = 0; i < listOfVms.length; i++) {
+                    var vmName = listOfVms[i]["name"];
+                    var resourceGroupName = this._taskParameters.resourceGroupName;
+                    var extensionParameters = this._formExtensionParameters(listOfVms[i], operationParameters.operation);
+                    var extensionName = extensionParameters["extensionName"];
+                    var parameters = extensionParameters["parameters"];
+                    var callback = this._getPostOperationCallBack(operationParameters, extensionName, vmName);
+                    var createOrUpdateCallback = this._getCreateOrUpdateCallback(operationParameters, extensionName, vmName, parameters, callback);
+                    this._computeClient.virtualMachines.get(resourceGroupName, vmName, { expand: 'instanceView' }, createOrUpdateCallback);
+                }
+            });
+            return operationParameters.deferred.promise;
+        }
+        catch (exception) {
+            this.log(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            tl.setResult(tl.TaskResult.Failed, tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            operationParameters.deferred.resolve(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+        }
+    }
+
+    public deleteMGExtensionRG() {
+        try {
+            var operationParameters = new OperationParameters("uninstallation");
+            operationParameters.deferred = Q.defer<string>();
+            var listOfVmsPromise = this._azureUtils.getVMDetails();
+            listOfVmsPromise.then((listOfVms) => {
+                operationParameters.vmCount = listOfVms.length;
+                if (operationParameters.vmCount === 0) {
+                    operationParameters.deferred.resolve("");
+                }
+                for (var i = 0; i < listOfVms.length; i++) {
+                    var vmName = listOfVms[i]["name"];
+                    var resourceGroupName = this._taskParameters.resourceGroupName;
+                    var extensionParameters = this._formExtensionParameters(listOfVms[i], operationParameters.operation);
+                    var extensionName = extensionParameters["extensionName"];
+                    var callback = this._getPostOperationCallBack(operationParameters, extensionName, vmName);
+                    var extensionDeletionCallback = this._getExtensionDeletionCallback(operationParameters, resourceGroupName, extensionName, vmName, callback);
```

Comment:
Not entirely, because the two methods use on promise for all vms and one promise per vm respectively. There is some common functionality between the two, though, which I have put in a separate function